sources/zafira-services/src/main/java/com/qaprosoft/zafira/service/WidgetService.java

```diff
@@ -79,15 +94,95 @@ public Widget getWidgetById(long id) {
 
     @Transactional(rollbackFor = Exception.class)
     public Widget updateWidget(Widget widget) {
+        if (widget.getWidgetTemplate() != null) {
+            prepareWidgetTemplate(widget);
+        }
         widgetMapper.updateWidget(widget);
         return widget;
     }
 
+    private WidgetTemplate prepareWidgetTemplate(Widget widget) {
+        long templateId = widget.getWidgetTemplate().getId();
+        WidgetTemplate widgetTemplate = widgetTemplateService.getNotNullWidgetTemplateById(templateId);
+        widgetTemplateService.clearRedundantParamsValues(widgetTemplate);
+        widget.setWidgetTemplate(widgetTemplate);
+        return widgetTemplate;
+    }
+
     @Transactional(rollbackFor = Exception.class)
     public void deleteWidgetById(Long id) {
         widgetMapper.deleteWidgetById(id);
     }
 
+    @Transactional(readOnly = true)
+    public List<Map<String, Object>> getQueryResults(boolean stackTraceRequired, Map<String, Object> params, Long templateId, Long userId, String userName) {
+        WidgetTemplate widgetTemplate = widgetTemplateService.getNotNullWidgetTemplateById(templateId);
+        List<Map<String, Object>> resultList;
+        try {
+            Map<WidgetService.DefaultParam, Object> additionalParams = new HashMap<>();
+            additionalParams.put(WidgetService.DefaultParam.CURRENT_USER_NAME, userName);
+            additionalParams.put(WidgetService.DefaultParam.CURRENT_USER_ID, userId);
+            resultList = executeSQL(widgetTemplate.getSql(), params, additionalParams, true);
+        } catch (Exception e) {
+            if (stackTraceRequired) {
+                resultList = List.of(Map.of("Check your query", ExceptionUtils.getFullStackTrace(e)));
+            } else {
+                // wrap whatever error is thrown
+                throw new ProcessingException(WIDGET_QUERY_EXECUTION_ERROR, e.getMessage(), e);
+            }
+        }
+        return resultList;
+    }
+
+    /**
+        Used for old widget versions and table widgets.
+        Someday we'll remove echarts library completely and refactor this method.
+     */
+
+    @Transactional(readOnly = true)
+    public List<Map<String, Object>> getQueryResultObsolete(List<String> projects,
+                                                            String currentUserId,
+                                                            String dashboardName,
+                                                            boolean stackTraceRequired,
+                                                            String query,
+                                                            List<Attribute> attributes,
+                                                            Long userId,
+                                                            String userName) {
+        List<Map<String, Object>> resultList;
+        try {
+            query = applyAttributes(attributes, query);
+            query = query
```

Comment:
Let's move this to private method with proper name. (i mean all query.replaceAll ...)